src/psse_export.jl

```diff
@@ -0,0 +1,599 @@
+const PSSE_EXPORT_SUPPORTED_VERSIONS = [:v33]  # TODO add :v34
+const PSSE_DEFAULT = ""  # Used below in cases where we want to insert an empty field to signify the PSSE default
+const PSSE_BUS_TYPE_MAP = Dict(
+    PSY.ACBusTypes.PQ => 1,
+    PSY.ACBusTypes.PV => 2,
+    PSY.ACBusTypes.REF => 3,
+    PSY.ACBusTypes.SLACK => 3,
+    PSY.ACBusTypes.ISOLATED => 4,
+)
+
+# TODO consider adding this to IS
+"""
+A context manager similar to `Logging.with_logger` that sets the system's units to the given
+value, executes the function, then sets them back. Suppresses logging below `Warn` from
+internal calls to `set_units_base_system!`.
+"""
+function with_units(f::Function, sys::System, units::Union{PSY.UnitSystem, String})
+    old_units = PSY.get_units_base(sys)
+    Logging.with_logger(Logging.SimpleLogger(Logging.Warn)) do
+        PSY.set_units_base_system!(sys, units)
+    end
+    try
+        f()
+    finally
+        Logging.with_logger(Logging.SimpleLogger(Logging.Warn)) do
+            PSY.set_units_base_system!(sys, old_units)
+        end
+    end
+end
+
+"""
+Structure to perform an export from a Sienna System, plus optional updates from
+`PowerFlowData`, to the PSS/E format. Construct from a `System` and a PSS/E version, update
+using `update_exporter` with any new data as relevant, and perform the export with
+`write_export`.
+
+# Arguments:
+  - `base_system::PSY.System`: the system to be exported. Later updates may change power
+    flow-related values but may not fundamentally alter the system
+  - `psse_version::Symbol`: the version of PSS/E to target, must be one of
+    `PSSE_EXPORT_SUPPORTED_VERSIONS`
+"""
+mutable struct PSSEExporter
+    # Internal fields are very much subject to change as I iterate on the best way to do
+    # this! For instance, the final version will almost certainly not store an entire System
+    system::PSY.System
+    psse_version::Symbol
+
+    function PSSEExporter(base_system::PSY.System, psse_version::Symbol)
+        (psse_version in PSSE_EXPORT_SUPPORTED_VERSIONS) ||
+            throw(
+                ArgumentError(
+                    "PSS/E version $psse_version is not supported, must be one of $PSSE_EXPORT_SUPPORTED_VERSIONS",
+                ),
+            )
+        system = deepcopy(base_system)
+        new(system, psse_version)
+    end
+end
+
+function _validate_same_system(sys1::PSY.System, sys2::PSY.System)
+    return IS.get_uuid(PSY.get_internal(sys1)) == IS.get_uuid(PSY.get_internal(sys2))
+end
+
+"""
+Update the `PSSEExporter` with new `data`.
+
+# Arguments:
+  - `exporter::PSSEExporter`: the exporter to update
+  - `data::PSY.PowerFlowData`: the new data. Must correspond to the `System` with which the
+    exporter was constructor
+"""
+function update_exporter!(exporter::PSSEExporter, data::PowerFlowData)
```

Comment:
For reference since we discussed it already. Relying on deep copies can be problematic for systems with large time series data sets.

Reply:
Yes — will change this in Part 2